### ISSUE TITLE
Add `mask_rom` padding and report utilization at build time

### ISSUE THREAD

**Issue by moidx** [lowRISC/opentitan#5771]
This issue is a follow up of lowrisc/opentitan#5694. 

We should add mask_rom padding to the expected ROM size, and report utilization as part of the build flow.

We should also discuss what type of padding to use to mitigate against any potential fault attacks, and/or what isolation configuration to use out of reset. In any case.

`/home/moidx/opentitan/hw/top_earlgrey/util/vivado_hook_opt_design_post.tcl` should also be updated to check for the right number of FPGA BRAM blocks. This is contingent on `util/fpga/splice_nexysvideo.sh` support.

**Comment by moidx** [comment]
Added lowrisc/opentitan#5820 to track the implementation of binary size tracking integrated into the build system.

**Comment by moidx** [comment]
@mcy @cfrantz

**Comment by rswarbrick** [comment]
Possibly wait on the padding until the rom_ctrl stuff is merged? Since we want to avoid padding with (and "encrypting") zeros, my scrambling script currently takes the ELF and adds a pseudo-random sequence to fill in the gaps. So, basically, I might have done the work already :-)

**Comment by cdgori** [comment]
We should probably have a quick cryptographic review of the padding scheme once it's done.  There's lots of weird edge cases there.

**Comment by cdgori** [comment]
Thanks for that detailed breakdown @rswarbrick !

I think the key parts to review then are the 2nd and 4th bullet, since that's where the actual padding insertion happens (either at the 32-bit-word level or at the chunk level).  I am not nearly cryptographically qualified enough to review the padding though, since there are many subtle ways it can go wrong, maybe @moidx knows someone whom we could "volunteer" for this task?

**Comment by moidx** [comment]
Hi @johannheyszl, can you triage this from a security perspective and determine if we need to get others involved? We should aim to have discussion points for this for the next security WG so that we can close sign off D3 for the rom_ctrl block. Thanks!

**Comment by johannheyszl** [comment]
Relevant info IMO:
- ROM uses reduced-round PRINCE in CTR mode with additional diffusion within words (see hw/ip/rom_ctrl/rtl/rom_ctrl_scrambled_rom.sv which refers to hw/ip/prim/rtl/prim_ram_1p_scr.sv)
- ROM content is padded as described above (partly with PRNG generated sequence)
- Confidentiality of ROM is not a goal AFAIK
- Integrity seems protected with a cSHAKE-256 hash by comparison to a value in ROM

Open question is regarding padding / crypto / sec:
- Is the padding words out with zeros ok? I think yes. @jadephilipoom pls WDYT?
- Is the PRNG-based pading ok? I think it could probably be zeros as well. @jadephilipoom pls WDYT?

**Comment by jadephilipoom** [comment]
I think there should be no issue with padding using zeroes in this case. Since a lot of the ROM is based on open-source code, the attacker already has a lot of known plaintext to work with; I don't see why adding a few extra zeroes would cause a problem that the known code doesn't already cause. But maybe I've missed a reason why these would be an issue? @cdgori , was there another reason for avoiding zero-padding other than creating known plaintext?

It's true that a lot of known plaintext weakens PRINCE, but you need a pretty huge number (at least with all the attacks I could find). For example, [this paper ](http://research.ics.aalto.fi/publications/bibdb2012/public_pdfs/FSE2013.pdf)claims key recovery given 2^58 known plaintexts and 2^72 encryptions. According to [these NIST slides](https://csrc.nist.gov/csrc/media/events/lightweight-cryptography-workshop-2015/documents/presentations/session2-rechberger.pdf), the security is about 126-n bits when the attacker knows 2^n plaintext/ciphertext pairs. An upper bound for 2^n in our case is <size of flash>/ <block size = 64 bits>. Given that this is to make the attacker's life difficult (scrambling rather than encryption), I think this is unlikely to reduce the bound so much that it won't be at least annoying. Does anyone happen to know the size of flash, or a ballpark size of a ROM image?

**Comment by johannheyszl** [comment]
@jadephilipoom thank you and agree.

Mem sizes are listed here https://docs.opentitan.org/hw/top_earlgrey/doc/
32kB ROM means 4k = 2^12 plaintext/ciphertext pairs.
This should mean 126-12=114 bits of security which should be sufficient here (attacker has to extract image first).

@moidx and @cdgori do you remember any other reason for avoiding zeros that we are missing?

**Comment by moidx** [comment]
This issue was probably added before we agreed on using cSHAKE for the digest calculation. Given that the attacker already has access to the ROM source code, I can only think of the following attack scenarios: 

1. Recovering the scrambling key to be able to identify sections in the code that could be targeted in scrambled form. This is mitigated by the cSHAKE measurement. 
2. Manipulation of the target cSHAKE digest comparator value. This is also stored in ROM if I am not mistaken. I am not sure if this value is protected by scrambling. I suspect that regardless of the value being scrambled, this would be a high effort attack, considering the attack identification difficulty as mentioned by @johannheyszl in the previous comment.

I am on the fence regarding using PRNG data versus 0's for padding in this case. My concern is that the ePMP default settings probably make some of this padding data executable memory? If so, I'd prefer it to be an invalid instruction versus running into the risk of being able to decode into a malicious instruction.  However, I don't have a way to quantify how feasible this attack is. One simple mitigation would be to tweak the ePMP default ROM execution window to solve this problem. The ROM itself could probably adjust the window if necessary.

I am resolving this issue based on the feedback from @jadephilipoom and @johannheyszl. @msfschaffner please let us know if there are any additional opens here.

**Comment by msfschaffner** [comment]
To your questions @moidx:
- the hash value itself is stored in unscrambled form and read verbatim for the comparison.
- changing the PRNG data to zero should be an easy change in the scrambling script. if that is preferred, I can change this right away.
https://github.com/lowRISC/opentitan/blob/1447b4853447dffbc0193b5aab27d96012f950ff/hw/ip/rom_ctrl/util/mem.py#L279-L321

**Comment by moidx** [comment]
@alphan wdyt?

**Comment by alphan** [comment]
> My concern is that the ePMP default settings probably make some of this padding data executable memory? I

Correct. Entire ROM region is LRX at [reset](https://github.com/lowRISC/opentitan/blob/aacbe3f39b498bd0e7aef611ab925cca5963c181/hw/ip/rv_core_ibex/rtl/ibex_pmp_reset.svh#L37), but reconfigured soon after [here](https://github.com/lowRISC/opentitan/blob/master/sw/device/silicon_creator/rom/rom_epmp_init.S).

Revisiting reset values definitely sounds like a good idea since it would improve the security posture.

**Comment by alphan** [comment]
We could also mask this PRNG data to make it invalid as we discussed offline.

**Comment by alphan** [comment]
From 2022-08-18 security meeting:
- [ ] @msfschaffner could you please update the script to pad with zeros?
- [ ] @alphan will look into shrinking the ePMP reset config.
